// tiktok-live-connector/dist/lib/ws/lib/ws-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const websocket_1 = require("websocket");
const utilities_1 = require("../../../lib/utilities");
const config_1 = __importDefault(require("../../../lib/config"));
const types_1 = require("../../../types");
const textEncoder = new TextEncoder();
class TikTokWsClient extends websocket_1.client {
    webSocketParams;
    webSocketPingIntervalMs;
    connection;
    pingInterval;
    wsHeaders;
    wsUrlWithParams;
    constructor(wsUrl, cookieJar, webSocketParams, webSocketHeaders, webSocketOptions, webSocketPingIntervalMs = 10000) {
        super();
        this.webSocketParams = webSocketParams;
        this.webSocketPingIntervalMs = webSocketPingIntervalMs;
        this.pingInterval = null;
        this.connection = null;
        this.wsUrlWithParams = `${wsUrl}?${new URLSearchParams(this.webSocketParams)}${config_1.default.DEFAULT_WS_CLIENT_PARAMS_APPEND_PARAMETER}`;
        this.wsHeaders = { Cookie: cookieJar.getCookieString(), ...(webSocketHeaders || {}) };
        this.on('connect', this.onConnect.bind(this));
        this.connect(this.wsUrlWithParams, '', `https://${config_1.default.TIKTOK_HOST_WEB}`, this.wsHeaders, webSocketOptions);
    }
    onConnect(wsConnection) {
        this.sendHeartbeat();
        this.connection = wsConnection;
        this.pingInterval = setInterval(() => this.sendHeartbeat(), this.webSocketPingIntervalMs);
        this.connection.on('message', this.onMessage.bind(this));
        this.connection.on('close', this.onDisconnect.bind(this));
    }
    /**
     * Send a message to the WebSocket server
     * @param data The message to send
     * @returns True if the message was sent, false otherwise
     */
    sendBytes(data) {
        if (this.connection) {
            this.connection.sendBytes(Buffer.from(data));
            return true;
        }
        return false;
    }
    onDisconnect() {
        clearInterval(this.pingInterval);
        this.pingInterval = null;
        this.connection = null;
        this.emit('close');
    }
    /**
     * Handle incoming messages
     * @param message The incoming WebSocket message
     * @protected
     */
    async onMessage(message) {
        // Emit WebSocket data
        this.emit('webSocketData', message);
        // If the message is not binary, emit an unknown response
        if (message.type !== 'binary') {
            return this.emit('unknownResponse', message);
        }
        //  If the message is binary, decode it
        try {
            const decodedContainer = await (0, utilities_1.deserializeWebSocketMessage)(message.binaryData);
            // If the message has a decoded protoMessageFetchResult, emit it
            if (decodedContainer.protoMessageFetchResult) {
                // If it needs an ack, send the ack
                if (decodedContainer.protoMessageFetchResult.needsAck) {
                    this.sendAck(decodedContainer);
                }
                this.emit('protoMessageFetchResult', decodedContainer.protoMessageFetchResult);
            }
            // If it's a room enter, emit
            if (decodedContainer.payloadType === 'im_enter_room_resp') {
                this.emit('imEnteredRoom', decodedContainer);
            }
        }
        catch (err) {
            this.emit('messageDecodingFailed', err);
        }
    }
    /**
     * Static Keep-Alive ping
     */
    sendHeartbeat() {
        const { room_id } = this.webSocketParams;
        // Create the heartbeat
        const hb = types_1.HeartbeatMessage.encode({ roomId: room_id });
        // Wrap it in the WebcastPushFrame
        const webcastPushFrame = (0, utilities_1.createBaseWebcastPushFrame)({
            payloadEncoding: 'pb',
            payloadType: 'hb',
            payload: hb.finish(),
            service: undefined,
            method: undefined,
            headers: {}
        });
        this.sendBytes(Buffer.from(webcastPushFrame.finish()));
    }
    /**
     * EXPERIMENTAL: Switch to a different TikTok LIVE room while connected to the WebSocket
     * @param roomId The room ID to switch to
     */
    switchRooms(roomId) {
        const imEnterRoomMessage = types_1.WebcastImEnterRoomMessage.encode({
            roomId: roomId,
            roomTag: '',
            liveRegion: '',
            liveId: '12',
            identity: 'audience',
            cursor: '',
            accountType: '0',
            enterUniqueId: '',
            filterWelcomeMsg: '0',
            isAnchorContinueKeepMsg: false
        });
        const webcastPushFrame = (0, utilities_1.createBaseWebcastPushFrame)({
            payloadEncoding: 'pb',
            payloadType: 'im_enter_room',
            payload: imEnterRoomMessage.finish()
        });
        this.sendBytes(Buffer.from(webcastPushFrame.finish()));
    }
    /**
     * Acknowledge the message was received
     */
    sendAck({ logId, protoMessageFetchResult: { internalExt } }) {
        // Always send an ACK for the message
        if (!logId) {
            return;
        }
        const webcastPushFrame = (0, utilities_1.createBaseWebcastPushFrame)({
            logId: logId,
            payloadEncoding: 'pb',
            payloadType: 'ack',
            payload: textEncoder.encode(internalExt)
        });
        this.sendBytes(Buffer.from(webcastPushFrame.finish()));
    }
    /**
     * Close the WebSocket connection
     */
    close() {
        return new Promise((resolve) => {
            this.once('close', () => resolve());
            // If connected, disconnect
            if (this.connection) {
                this.connection.close(1000);
            }
            // Otherwise immediately resolve
            else {
                resolve();
            }
        });
    }
}
exports.default = TikTokWsClient;
